nodes/arm_node.cpp

```diff
@@ -1,110 +1,247 @@
 #include <ros/ros.h>
+#include <ros/console.h>
+#include <ros/package.h>
+
+#include <actionlib/server/simple_action_server.h>
+
 #include <geometry_msgs/Point.h>
+
 #include <hebi_cpp_api_examples/TargetWaypoints.h>
+#include <hebi_cpp_api_examples/ArmMotionAction.h>
+
+#include "group_command.hpp"
 
 #include "robot/arm.hpp"
 #include "robot_model.hpp"
 
-#include <ros/console.h>
-#include <ros/package.h>
-
 namespace hebi {
-  namespace ros {
-
-    class ArmNode {
-    public:
-      ArmNode(arm::Arm& arm)
-        : arm_(arm)
-      { }
-
-      // "Jog" the target end effector location in (x,y,z) space, replanning
-      // smoothly to the new location
-      void offsetTargetCallback(geometry_msgs::Point data) {
-        
-        // Only update if target changes!
-        if (data.x == 0 && data.y == 0 && data.z == 0)
-          return;
-
-        // Initialize target from feedback as necessary
-        if (!isTargetInitialized()) {
-          auto pos = arm_.getKinematics().FK(arm_.getLastFeedback().getPosition());
-          target_xyz_.x() = pos.x();
-          target_xyz_.y() = pos.y();
-          target_xyz_.z() = pos.z();
-        }
-
-        // Update the target point
-        target_xyz_.x() += data.x / 100.0; // Converts to cm
-        target_xyz_.y() += data.y / 100.0; // Converts to cm
-        target_xyz_.z() += data.z / 100.0; // Converts to cm
-         
-        Eigen::VectorXd ik_result_joint_angles =
-          arm_.getKinematics().solveIK(arm_.getLastFeedback().getPosition(),
-          target_xyz_);
-
-        // Replan:
-        arm_.getTrajectory().replan(
-          ::ros::Time::now().toSec(),
-          arm_.getLastFeedback(),
-          ik_result_joint_angles);
+namespace ros {
+
+class ArmNode {
+public:
+  ArmNode(arm::Arm& arm) : arm_(arm) { }
+
+  // Replan a smooth joint trajectory from the current location through a
+  // series of cartesian waypoints.
+  // xyz positions should be a 3xn vector of target positions
+  void updateCartesianWaypoints(const Eigen::Matrix3Xd& xyz_positions,
```

Comment:
Awesome. Looks like this is a new overload of `updateCartesianWaypoints`, right? The biggest issue I have with the C++ API is that there are no public facing functions which allow you to specify a fixed size dimension in a matrix (_e.g._, here `Eigen::Matrix3Xd == Eigen::Matrix<double, 3, Dynamic>`)

Reply:
:+1: yeah -- I hadn't realized there were these overloads.  We should totally use these on the gyros/accels if we're not yet.

Reply:
Basically, this is a helper method that all the other message/action callbacks reference -- I commented this more clearly and made it private